### PR TITLE
Revisit UnionContract design

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Key changes:
- Rename to UnionContract
- Add `IUnionContract` marker interface. All union contracts must now implement this as a safety measure.
- Allow Nullary union cases

### CODE REVIEW

File: src/TypeShape/Applications/UnionContract.fs [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+type IEncoder<'Format> =
+    /// 'Null' format value to be used for union cases without a payload
+    abstract Empty : 'Format
+    /// Encodes any value to the implementation format
+    abstract Encode<'T> : value:'T -> 'Format
+    /// Decodes format instance to any value
+    abstract Decode<'T> : fmt:'Format -> 'T
+
+/// Generic encoder that simply upcasts values to System.Object
+type CastEncoder() =
```
Comment (bartelink) [inline src/TypeShape/Applications/UnionContract.fs]: BoxEncoder?

File: src/TypeShape/Applications/UnionContract.fs [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    abstract Encode      : value:'Union -> EncodedUnion<'Format>
+    /// Decodes a formatted representation into a union instance. Can potentially throw FormatException
+    abstract Decode      : encodedUnion:EncodedUnion<'Format> -> 'Union
+    /// Decodes a formatted representation into a union instance. Does not throw exception on format mismatches
+    abstract TryDecode   : encodedUnion:EncodedUnion<'Format> -> 'Union option
+
+[<AutoOpen>]
+module private Impl =
+
+    type Config = { requireRecordPayloads : bool ; allowNullaryCases : bool }
```
Comment (bartelink) [inline src/TypeShape/Applications/UnionContract.fs]: can you put xmldoc on these ? (reason I ask is that I'm asking myself as I read: should those be mutually exclusive and/or a DU?)

File: src/TypeShape/Applications/UnionContract.fs [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+            String.concat ", " duplicates
+            |> sprintf "Union type '%O' defines the following duplicate case identifiers: %s" typeof<'Union>
+            |> invalidArg "Union"
+
+        let labelIndex = BinSearch labels
+
+        fun config encoder ->
+            match caseWithoutRecordPayload with
+            | Some case when config.requireRecordPayloads ->
+                sprintf "Union case '%O.%O' contains field which is not an F# record" typeof<'Union> case.CaseInfo.Name
```
Comment (bartelink) [inline src/TypeShape/Applications/UnionContract.fs]: which -> that ;)

File: src/TypeShape/Applications/UnionContract.fs [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+                    | tag -> caseDecoders.[tag] e.Payload
+                
+                member __.TryDecode e =
+                    match labelIndex.TryFindIndex e.CaseName with
+                    |  -1 -> None
+                    | tag -> caseDecoders.[tag] e.Payload |> Some }
+
+    type EncoderFactory<'Union, 'Format when 'Union :> IUnionContract> private () =
+        static let factory = lazy(mkUnionEncoder<'Union, 'Format>())
+        static member Create requireRecordFields (e : IEncoder<_>) = factory.Value requireRecordFields e
```
Comment (bartelink) [inline src/TypeShape/Applications/UnionContract.fs]: line break please

File: src/TypeShape/Applications/UnionContract.fs (hunk 2 above)
Comment (eiriktsarpalis) [reply src/TypeShape/Applications/UnionContract.fs]: These are private implementation types

Comment (bartelink) [reply src/TypeShape/Applications/UnionContract.fs]: doesn't mean it can't have a comment!

**Pull request merged**